third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -3724,40 +3935,118 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
+        // compute:
+        //     - the struct id of S
+        //     - (x_1, e_1), (x_2, e_2), ...
+        //     - arg_1, arg_2, ...
+        // such that the transformed code
+        //     { let x_1 = e_1; let x_2 = e_2; ...; pack(arg1, arg2, ...) }
+        // is equivalent.
+        let mut bindings = BTreeMap::new();
+        let mut args = BTreeMap::new();
+        let field_decls = self
+            .get_field_decls_for_pack_unpack(
+                &struct_entry,
+                &struct_name,
+                &struct_name_loc,
+                variant,
+            )?
+            .clone();
+        if let Some(fields) = fields {
+            self.check_missing_or_undeclared_fields(loc, struct_name, &field_decls, fields)?;
+            let in_order_fields = self.in_order_fields(&field_decls, fields);
+            for (_, name, (exp_idx, field_exp)) in fields.iter() {
+                let (def_idx, field_name, translated_field_exp) =
+                    self.translate_exp_field(&field_decls, name, &instantiation, field_exp);
+                if in_order_fields.contains(&def_idx) {
+                    args.insert(def_idx, translated_field_exp);
+                } else {
+                    // starts with $ for internal generated vars
+                    let var_name = self
+                        .symbol_pool()
+                        .make(&format!("${}", field_name.display(self.symbol_pool())));
+                    // the x_i to be used in the let bindings
+                    let var = Pattern::Var(translated_field_exp.node_id(), var_name);
+                    // the x_i to be used in the pack exp
+                    let arg = ExpData::LocalVar(translated_field_exp.node_id(), var_name);
+                    args.insert(def_idx, arg);
+                    bindings.insert(*exp_idx, (var, translated_field_exp));
+                }
+            }
+        } else if !field_decls.is_empty() {
+            self.error(
+                loc,
+                &format!(
+                    "no arguments provided for pack, expected {}",
+                    field_decls.len()
+                ),
+            )
+        }
+        let struct_id = struct_entry
+            .module_id
+            .qualified_inst(struct_entry.struct_id, instantiation);
+        let bindings = bindings
+            .into_iter()
+            .sorted_by_key(|(i, _)| *i)
+            .map(|(_, value)| value)
+            .collect_vec();
+        let args = args
+            .into_iter()
+            .sorted_by_key(|(i, _)| *i)
+            .map(|(_, value)| value)
+            .collect_vec();
+
+        let struct_ty = struct_id.to_type();
+        let struct_ty = self.check_type(loc, &struct_ty, expected_type, context);
+        let mut field_args = args.into_iter().map(|e| e.into_exp()).collect_vec();
+        if variant.is_none() && field_args.is_empty() {
+            // The move compiler inserts a dummy field with the value of false
```

Comment:
I am wondering whether zero-variant enum can be allowed?

Reply:
It is right now, there is no particular technical or semantic reason to forbid it. We can come back if we want to ban it.